File: node-ts/src/exceptions.ts
/**
 * Domain exceptions, mapped to HTTP status codes by the error handler.
 */

export class ParkingError extends Error {
  readonly statusCode: number;
  constructor(message: string, statusCode = 400) {
    super(message);
    this.name = this.constructor.name;
    this.statusCode = statusCode;
  }
}

export class NoSpotAvailable extends ParkingError {
  constructor(message: string) {
    super(message, 409);
  }
}

export class VehicleAlreadyParked extends ParkingError {
  constructor(message: string) {
    super(message, 409);
  }
}

export class TicketNotFound extends ParkingError {
  constructor(message: string) {
    super(message, 404);
  }
}

export class TicketAlreadyClosed extends ParkingError {
  constructor(message: string) {
    super(message, 409);
  }
}

File: app/exceptions.py
"""Domain exceptions, mapped to HTTP responses by the API layer."""

from __future__ import annotations


class ParkingError(Exception):
    """Base class for all domain errors."""

    status_code: int = 400


class NoSpotAvailable(ParkingError):
    status_code = 409


class VehicleAlreadyParked(ParkingError):
    status_code = 409


class TicketNotFound(ParkingError):
    status_code = 404


class TicketAlreadyClosed(ParkingError):
    status_code = 409


class InvalidConfiguration(ParkingError):
    status_code = 400
